test/test_allocation_domain.cpp

```diff
@@ -61,4 +61,246 @@ TEST_F(AllocationDomainTest, TransposedIntermediate_CUDA) {
+}
+
+// A global->global copy kernel converting NCHW memory format into NHWC, with a
+// 1d allocation domain in output.
+TEST_F(AllocationDomainTest, NCHW4d_To_NHWC1d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(4);
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  // [N, C, H, W]
+  tv1->reorder({{1, -1}});
+  // [N, H, W, C]
+  tv1->merge(0);
+  tv1->merge(0);
+  tv1->merge(0);
+  tv1->setAllocationDomain({tv1->axis(0)}, true);
+  // [N*H*W*C]
+  tv1->split(0, 128);
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, TIDx]
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  int n = 31, h = 64, w = 103, c = 21;
+
+  at::Tensor t0 = at::randn({n, c, h, w}, options);
+
+  FusionExecutor fe;
+  fe.compileFusion(fusion_ptr.get(), {t0});
+
+  auto cg_outputs = fe.runFusion({t0});
+
+  ASSERT_TRUE(cg_outputs[0].is_contiguous(at::MemoryFormat::ChannelsLast));
+
+  testValidate(&fusion, cg_outputs, {t0}, {t0}, __LINE__, __FILE__);
+}
+
+// A global->global copy kernel converting NCHW memory format into NHWC, with a
+// 2d allocation domain in output.
+TEST_F(AllocationDomainTest, NCHW4d_To_NHWC2d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(4);
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  // [N, C, H, W]
+  tv1->reorder({{1, -1}});
+  // [N, H, W, C]
+  tv1->merge(0);
+  tv1->merge(0);
+  tv1->merge(0);
+  // [N*H*W*C]
+  tv1->split(0, 128);
+  tv1->setAllocationDomain({tv1->axis(0), tv1->axis(1)}, true);
+  // [N*H*W*C/128, 128]
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, TIDx]
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  int n = 31, h = 64, w = 104, c = 21;
+
+  at::Tensor t0 = at::randn({n, c, h, w}, options);
+
+  FusionExecutor fe;
+  fe.compileFusion(fusion_ptr.get(), {t0});
+
+  auto cg_outputs = fe.runFusion({t0});
+
+  ASSERT_TRUE(cg_outputs[0].is_contiguous(at::MemoryFormat::ChannelsLast));
+
+  testValidate(&fusion, cg_outputs, {t0}, {t0}, __LINE__, __FILE__);
+}
+
+// Reshape and transpose a 3d tensor into an NHWC tensor with a 3d allocation
+// domain in fusion output.
+TEST_F(AllocationDomainTest, Tensor3d_To_NHWC3d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  int n1 = 31, n2 = 29, h = 64, w = 104, c = 21;
+
+  auto tv0 = makeContigTensor(3); // [N1, N2, H*W*C]
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  tv1->setAllocationDomain({tv1->axis(0), tv1->axis(1), tv1->axis(2)}, true);
```

Comment:
This one doesn't look clear to me. Is it `commitLeafToRFactor` that triggers the output stride order to change?

Basically I was under the impression that once we called `setAllocationDomain`, we have already pinned. the output to a certain stride order. (contiguous if we look at the format specified on the line above).
But looks like later scheduling did change how output turns is produced `at::MemoryFormat::ChannelsLast`.

So when we are defining the fusion, we'd want to dictate on output stride order. If the scheduler can change that, wouldn't it be error prone, basically a scheduler might accidentally alter the output format.

Reply:
IIUC, you're right that with `setAllocationDomain` that's the memory format. `commitLeafToRFactor`, which still feels a little weird to me, changes the logical shape of the tensor. I guess in this case it makes the tensor conform with the convention in PyTorch.

Reply:
> Basically I was under the impression that once we called `setAllocationDomain`, we have already pinned. the output to a certain stride order.

Yes, your understanding is correct.

The purpose of `commitLeafToRFactor` has nothing to do with memory layout. This function is part of fusion definition. And it only has something to do with tensor's semantics.

For example, initially, you created a tensor as `[I1, I2]`, then you scheduled it to get a leaf domain `[I1*I2]`, after you call `commitLeafToRFactor`, then the leaf domain of this tensor will become rfactor domain, which means that the semantics of this tensor has changed. This function is mostly used for defining "implicit view", i.e., a view that does not happen at a `ViewOp`. For example, you might have
```
T1 = sin(T0)

T0:
root domain: [I1, I2]

T1:
root domain: [I1, I2]
rFactor domain: [I1*I2]
```
Then `T1` has an implicit view, because the view happens at an `UnaryOp` with type `sin`, instead of at a `ViewOp`. 

The concept of implicit-view is introduced recently, which I used for adding NN support for Volta matmul. Basically, what I needed is not implicit-view, but implicit transpose. The output of `MmaOp` needs to have an implicit transpose to change the tensor semantics from `[N,M]` to `[M,N]`.

Today, and in this test, `commitLeafToRFactor` is only used to define fusions. And I am not expecting schedulers to use it. In the future, I might want to let schedulers to use it, but there will be rules:
1. Schedulers can only use this function on intermediate tensors, not inputs/outputs
2. When this function is called, all its consumers needs to be transformed to revert the effect of this function. For example, if you have `T0->T1->T2`, and you did an implicit view on `T1`, the you will have to create an implicit "unview" on `T2` to so that the semantics of `T2` does not change.

Reply:
> Schedulers can only use this function on intermediate tensors, not inputs/outputs

But that's exactly what this test violates, which made me feel a little uncomfortable, although this is just a test.

Reply:
Hmm, what does this test violate? I would consider the lines below `commitLeafToRFactor` as "scheduler", and lines before and at `commitLeafToRFactor` as fusion definition. So nothing is violated.🤓 Maybe it would make us feel more comfortable if I move the `fusion.addOutput(tv1);` after the `commitLeafToRFactor`? Sometimes I also feel that I have a mental barrier that a fusion definition should end with `addOutput`s, and anything after that are considered scheduling.

Reply:
`tv1` is an output, yet the test uses `commiteLeafToRFactor` with it.

Reply:
I am not saying that `commiteLeafToRFactor` should never be used on fusion outputs at any time. I was saying that `commiteLeafToRFactor` shouldn't be applied to fusion outputs **by schedulers**. And in this test, the purpose of `commiteLeafToRFactor` is to define the fusion, not to schedule it. It is totally fine to apply `commiteLeafToRFactor` to any tensor, including fusion inputs and outputs, for the purpose of defining a fusion.

Reply:
OK, I didn't consider it's part of fusion definition.